### ISSUE TITLE
Tests do not compile with Scala 2.11

### ISSUE THREAD

**Issue by propensive** [propensive/magnolia#78]
Looks like @joroKr21's stack reworking introduced something, which appears to affect only Scala 2.11, and currently stops four tests from compiling... I've seen error messages suggestion something relating to lazy vals.

**Comment by joroKr21** [comment]
I don't know about lazy vals, but one thing I noticed for sure in #73 is that `DirectlyReentrantException` always triggers on recursive types, but then we recover. That happens because we first try the regular implicit resolution and we always find our own macro and then we fall back to direct recursive derivation. That may or may not have happened before but apart from causing #73 it must be horribly inefficient so there is even room for speedup! I'm not sure how to fix it yet though.

**Comment by propensive** [comment]
The `DirectlyReentrantException` hack was very much part of the original design, and it effectively uses implicit search to check whether there is a non-Magolia implicit in scope for that particular search which should take precedence. I don't know of a better way to do this... It's only an exception in that we want a nested macro invocation to fail so that we can search for other suitable implicits that would otherwise be shadowed...

I'm just investigating some more, though... :)

**Comment by propensive** [comment]
Here are the errors, by the way;
```
[info] Compiling 1 Scala source to /home/jpretty/dev/propensive.com/magnolia/tests/target/scala-2.11/classes ...
[error] /home/jpretty/dev/propensive.com/magnolia/tests/src/main/scala/tests.scala:130:17: could not find implicit value for parameter e: magnolia.examples.Show[String,magnolia.tests.FruitBasket]
[error]       implicitly[Show[String, FruitBasket]].show(FruitBasket(Fruit("apple"), Fruit("banana")))
[error]                 ^
[error] /home/jpretty/dev/propensive.com/magnolia/tests/src/main/scala/tests.scala:237:15: not found: value companyTypeclass$macro$255$lzy
[error]       Show.gen[Politician[String]].show(Corrupt("wall", Seq(Company("Alice Inc"))))
[error]               ^
[error] /home/jpretty/dev/propensive.com/magnolia/tests/src/main/scala/tests.scala:333:15: not found: value companyTypeclass$macro$352$lzy
[error]       Show.gen[Portfolio].show(Portfolio(Company("Alice Inc"), Company("Bob & Co")))
[error]               ^
[error] /home/jpretty/dev/propensive.com/magnolia/tests/src/main/scala/tests.scala:381:15: not found: value pathTypeclass$macro$425$lzy
[error]       Show.gen[Path[String]].show(OffRoad(Some(Crossroad(Destination("A"), Destination("B")))))
[error]               ^
[error] four errors found
```

**Comment by joroKr21** [comment]
I completely forgot about scala/bug#5466 (note the `$lzy` suffix)

**Comment by joroKr21** [comment]
Also, it seems that currently `crossScalaVersions in tests` is only 2.12

**Comment by propensive** [comment]
Ah, does that mean we can get it working again on Scala 2.11 with some targetted untypechecking?

**Comment by joroKr21** [comment]
Rather the opposite - the problems is that we use `c.untypecheck` on code that contains lazy vals.

**Comment by propensive** [comment]
Oh, I see. I can see some logic (I think) for the cases where it's happening...

I imagine that in an absolute worst-case scenario we could replace them with some dummy code, and rewrite that code as actual `lazy val`s when the macro finally produces output. But maybe you know a more straightforward workaround?

**Comment by joroKr21** [comment]
I don't know of any straightforward workaround. I'm not even sure what you're suggesting will work but you're welcome to try. In general the splicing of typed and untyped trees is an unsolved problem in scala.reflect. You can try to avoid calling `untypecheck` (it's only one place). Maybe `typingTransform` can be used? No easy solution I'm afraid.

**Comment by adamw** [comment]
Fixed by removing 2.11 support :)
